accepted/2021/preview-features/preview-features.md

```diff
@@ -0,0 +1,418 @@
+
+</Project>
+```
+
+### Assembly info generation
+
+When we generate the `AssemblyInfo.cs`, we'll generate an assembly level
+attribute (the attribute and the meaning is discussed [later in this
+document](#api-analyzer)).
+
+For example, this project information:
+
+```xml
+<Project>
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <EnablePreviewFeatures>True</EnablePreviewFeatures>
+  </PropertyGroup>
+
+</Project>
+```
+
+Would result in this assembly-level attribute:
+
+```C#
+[assembly: RequiresPreviewFeatures]
+```
+
+This means you can look at a binary and tell whether it was built using preview
+features. This can be used for auditing and for detecting transitive
+dependencies ([discussed later](#api-analyzer)).
+
+### runtime.json
+
+Whether or not `EnablePreviewFeatures` was set to true for the application
+should be written into the `runtime.json` file such that the host/runtime can
+use it to make decisions. Alternatively, we could say this is encoded by the
+fact that the assembly the host considers the entry point has the
+`RequiresPreviewFeatures` assembly level attribute.
+
+### Compilation context
+
+When `EnablePreviewFeatures` is true, the `LangVersion` property should be set
+to `Preview`. This avoids customers having to turn on preview features for the
+language separately.
+
+In addition, the property `EnablePreviewFeatures` should be passed to the
+compilation context (akin to how `TargetFramework` is passed in today). An
+analyzer will use that to block use of APIs that are marked as preview as basing
+this off the language preview mode feels wrong.
+
+***Note:*** F# doesn't have this capability today.
+
+### API Analyzer
+
+We'll need an attribute to record whether or not the project was
+built with preview features turned on:
+
+```C#
+namespace System.Runtime.Versioning
+{
+    [AttributeUsage(AttributeTargets.Assembly |
+                    AttributeTargets.Module |
+                    AttributeTargets.Class |
+                    AttributeTargets.Struct |
+                    AttributeTargets.Enum |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Method |
+                    AttributeTargets.Property |
+                    AttributeTargets.Field |
+                    AttributeTargets.Event, AllowMultiple = true, Inherited = false)]
+    public sealed class RequiresPreviewFeaturesAttribute : Attribute
+    {
+        public PreviewAttribute();
+    }
+}
+```
+
+We'll ship a built-in analyzer that will report diagnostics when an API is being
+used that is marked `[RequiresPreviewFeatures]` but the consumer (member, type,
+module, assembly) isn't marked. The default severity is `error`.
+
+This analyzer will also validate that if any referenced assembly has applied
+`[RequiresPreviewFeatures]`, that the consuming assembly must also have this
+attribute applied.
+
+User code will typically only have this attribute applied at the assembly-level.
```

Comment:
What will the compiler do when the attribute is not applied at the assembly level?
Will it require that assembly is built with `<EnablePreviewFeatures>True</EnablePreviewFeatures>` ? 

If assembly A has a method M marked with `[RequiresPreviewFeatures]`, but no assembly level attribute, and it references assembly B that has the assembly level attribute, will the compiler warn?
If assembly A has `[RequiresPreviewFeatures]` assembly level attribute, and it references assembly B that has no assembly level attribute, but does have the attribute on a method M, will the compiler/analyzer flag it? what about if A is not using M?

etc

Reply:
I assume the answer is, the compiler errors when there is a call to another assembly that has the attribute on the called member, its type, or the assembly as  whole, and the calling method does not have the attribute on itself, its type, or the calling assembly as a whole.

I'm not sure about hte first question though.

Reply:
> What will the compiler do when the attribute is not applied at the assembly level?

I don't think the compiler/analyzer should do anything in this case. The SDK will generate the assembly level attribute based on the project setting, but this can be turned, which is what we will do to build the BCL. For obvious reason we don't System.Runtime as whole to have an assembly level attribute -- we only want this on specific types/members.

> If assembly A has a method M marked with [RequiresPreviewFeatures], but no assembly level attribute, and it references assembly B that has the assembly level attribute, will the compiler warn?

Yes. The assembly level attribute is meant to protect the user that just uses the project level setting. Advanced users (like us) can turn off the generation of the assembly level attribute and selectively apply this attribute to particular parts of the library. In this case it will only warn users that uses those APIs.

Reply:
I have added a scenario to capture this.